### PR TITLE
Allow configuration of blockchain's interpreter runtime

### PR DESCRIPTION
**Pull request by m-Peter**
Work towards: https://github.com/onflow/cadence-tools/pull/210

## Description

Also expose a public method for creating script environments based on blockchain's ledger state.
These utility functions facilitate the usage of emulator's blockchain, as a library for 3rd party tools (e.g. `cadence-tools/test`).

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the GitHub PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: emulator/blockchain.go [hunk 1]
```diff
@@ -555,8 +567,12 @@ func configureFVM(blockchain *Blockchain, conf config, blocks *blocks) (*fvm.Vir
 		CapabilityControllersEnabled: true,
 		CoverageReport:               conf.CoverageReport,
 	}
+	rt := conf.runtime
+	if conf.runtime == nil {
```
Comment (turbolent) [inline emulator/blockchain.go:571]:
```suggestion
	if rt == nil {
```

File: emulator/blockchain.go [hunk 2]
```diff
@@ -555,8 +567,12 @@ func configureFVM(blockchain *Blockchain, conf config, blocks *blocks) (*fvm.Vir
...
+		rt = runtime.NewInterpreterRuntime(config)
```
Comment (turbolent) [inline emulator/blockchain.go:572]: Unrelated: maybe rename `config` to `runtimeConfig` – it's a bit confusing that there's both a `conf` and a `config` in the function

File: emulator/blockchain.go [hunk 3]
```diff
@@ -1629,6 +1645,16 @@ func (b *Blockchain) SetClock(clock Clock) {
 	b.pendingBlock.SetClock(clock)
 }
 
+// NewScriptEnvironment returns an environment.Environment by
+// using as a storage snapshot the blockchain's ledger state.
+// Useful for tools that use the emulator's blockchain as a library.
+func (b *Blockchain) NewScriptEnvironment() environment.Environment {
+	return environment.NewScriptEnvironmentFromStorageSnapshot(
+		b.vmCtx.EnvironmentParams,
+		b.pendingBlock.ledgerState.NewChild(),
```
Comment (turbolent) [inline emulator/blockchain.go:1654]: Were these taken/copied from another place? Just curious, don't really know if that's how `NewScriptEnvironmentFromStorageSnapshot` is supposed to be initialized

**Review by turbolent — COMMENTED** [review]
Nice!

File: emulator/blockchain.go (hunk 2 above)
Comment (m-Peter) [reply emulator/blockchain.go]: Good point :+1: 
Fixed in https://github.com/onflow/flow-emulator/pull/476/commits/a349e2bb4826f44a787ee4f514f3023e65c42799

File: emulator/blockchain.go (hunk 1 above)
Comment (m-Peter) [reply emulator/blockchain.go]: Good point :+1: 
Fixed in https://github.com/onflow/flow-emulator/pull/476/commits/27408cfe3291fef240233369708d11933d1a33d8

File: emulator/blockchain.go (hunk 3 above)
Comment (m-Peter) [reply emulator/blockchain.go:1654]: Yes, I got them from the `Blockchain.GetSourceFile` method, so I have updated it to use this new helper function in: https://github.com/onflow/flow-emulator/pull/476/commits/e300dd2799807742587424072a915af36f44aecc

**Comment by m-Peter** [comment]
I also came across a bug, which I have fixed here: https://github.com/onflow/flow-emulator/pull/476/commits/cd45777151b9db004e2b873f5c5e0e191787c8d2

File: emulator/blockchain.go [hunk 4]
```diff
@@ -775,10 +791,12 @@ func configureTransactionValidator(conf config, blocks *blocks) *access.Transact
 }
 
 func (b *Blockchain) newFVMContextFromHeader(header *flowgo.Header) fvm.Context {
```
Comment (turbolent) [inline emulator/blockchain.go:793]: Maybe rename the function to indicate that the function not only constructs a new value (`new...`), but also sets it as the active context

**Review by turbolent — APPROVED** [review]

File: emulator/blockchain.go (hunk 4 above)
Comment (m-Peter) [reply emulator/blockchain.go:793]: Ah, that's a good point :+1: .
Maybe something like: `setFVMContextFromHeader` ?

**Comment by janezpodhostnik** [comment]
Is the injection via `NewCustomReusableCadenceRuntimePool` with the parameter `newCustomRuntime CadenceRuntimeConstructor` not usable? Not sure we need both.

**Comment by m-Peter** [comment]
@janezpodhostnik Very good point :clap:
I checked again with the companion PR that depends on this change (https://github.com/onflow/cadence-tools/pull/210) and the `WithRuntime` function is not actually needed. So I have removed it entirely.

**Review by janezpodhostnik — APPROVED** [review]

**Review by SupunS — APPROVED** [review]

**Pull request merged**